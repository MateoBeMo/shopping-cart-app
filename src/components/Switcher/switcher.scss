@import "styles/styles-utils.scss";

.switcher__wrapper{
    .switcher__switch {
        display: inline-block;
        height: 16px;
        position: relative;
        width: 32px;

        :before {
          @include themify($themes) {
            background-color: themed("switcherCircle");
          }
          bottom: 0px;
          content: "";
          height: 16px;
          left: 0px;
          position: absolute;
          transition: .4s;
          width: 16px;
        }
        
        input {
          display:none;
          &:checked{
            +.switcher__slider{
              @include themify($themes) {
                background-color: themed("switcher");
              }
              &:before{
                transform: translateX(16px);
              }
            }       
          }
        }
        
        .switcher__slider {
            background-color: #ccc;
            bottom: 0;
            cursor: pointer;
            left: 0;
            position: absolute;
            right: 0;
            top: 0;
            transition: .4s;
            &.switcher__round{
              border-radius: 24px;
              &:before {
                border-radius: 50%;
              }
            }
        }
    } 
}