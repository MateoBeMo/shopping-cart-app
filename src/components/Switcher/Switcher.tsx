import React, { ChangeEvent } from 'react';
/**Libraries */
/**Components */

/**Styles */
import './switcher.scss';

type SwitcherProps = {
    onChange: (event: ChangeEvent<HTMLInputElement>) => void,
    checked: boolean,
    switcherName: string
}

const Switcher = ({ onChange, checked, switcherName }: SwitcherProps) => {
    return (
        <div className="switcher switcher__wrapper">
            <label aria-label={switcherName} className="switcher__switch" htmlFor={switcherName}>
                <input
                    type="checkbox"
                    id={switcherName}
                    onChange={onChange}
                    checked={checked}
                    data-testid="switcher__input__checkbox-test"
                />
                <div className="switcher__slider switcher__round"></div>
            </label>
        </div>
    );
};

export default Switcher;
