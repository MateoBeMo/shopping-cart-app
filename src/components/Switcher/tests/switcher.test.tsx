import React, { ChangeEvent } from 'react';
/**Components */
import Switcher from '../Switcher';
/**Testing library */
import { render, fireEvent, cleanup } from '@testing-library/react';
import { axe } from 'jest-axe';

const onChangeMock = jest.fn();

type SwitcherProps = {
    onChange: (event: ChangeEvent<HTMLInputElement>) => void;
    checked: boolean;
    switcherName: string;
};

const renderSwitcher = (props: Partial<SwitcherProps> = {}) => {
    const defaultProps: SwitcherProps = {
        onChange: onChangeMock,
        checked: true,
        switcherName: 'switcher-test',
    };
    return render(<Switcher {...defaultProps} {...props} />);
};

describe('<Switcher>', () => {
    afterEach(() => {
        cleanup();
    });

    test('Snapshot renders => it should rendered with the provided props', () => {
        const { container } = renderSwitcher();

        expect(container).toMatchSnapshot();
    });

    test('It should fire onChange event on click over element', () => {
        const { getByTestId } = renderSwitcher();

        const inputCheckbox = getByTestId('switcher__input__checkbox-test');

        fireEvent.click(inputCheckbox);

        expect(onChangeMock).toHaveBeenCalled();
        expect(onChangeMock).toHaveBeenCalledTimes(1);
    });

    test('It should get checked prop', () => {
        const { getByTestId } = renderSwitcher();

        const checkbox = getByTestId('switcher__input__checkbox-test');

        expect(checkbox).toHaveAttribute('checked', '');
    });

    test('should not have basic accessibility issues', async () => {
        const { container } = renderSwitcher();
        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });
});
