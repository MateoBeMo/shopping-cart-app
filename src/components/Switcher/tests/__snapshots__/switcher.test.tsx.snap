// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switcher> Snapshot renders => it should rendered with the provided props 1`] = `
<div>
  <div
    class="switcher switcher__wrapper"
  >
    <label
      aria-label="switcher-test"
      class="switcher__switch"
      for="switcher-test"
    >
      <input
        checked=""
        data-testid="switcher__input__checkbox-test"
        id="switcher-test"
        type="checkbox"
      />
      <div
        class="switcher__slider switcher__round"
      />
    </label>
  </div>
</div>
`;
